egs/wsj/s5/steps/data/reverberate_data_dir.py

```diff
@@ -0,0 +1,470 @@
+def PickItemWithProbability(x):
+   if isinstance(x, dict):
+     plist = list(set(x.values()))
+   else:
+     plist = x
+   total_p = sum(item.probability for item in plist)
+   p = random.uniform(0, total_p)
+   accumulate_p = 0
+   for item in plist:
+      if accumulate_p + item.probability >= p:
+         return item
+      accumulate_p += item.probability
+   assert False, "Shouldn't get here as the accumulated probability should always equal to 1"
+
+
+# This function parses a file and pack the data into a dictionary
+# It is useful for parsing file like wav.scp, utt2spk, text...etc
+def ParseFileToDict(file, assert2fields = False, value_processor = None):
+    if value_processor is None:
+        value_processor = lambda x: x[0]
+
+    dict = {}
+    for line in open(file, 'r'):
+        parts = line.split()
+        if assert2fields:
+            assert(len(parts) == 2)
+
+        dict[parts[0]] = value_processor(parts[1:])
+    return dict
+
+# This function creates a file and write the content of a dictionary into it
+def WriteDictToFile(dict, file_name):
+    file = open(file_name, 'w')
+    keys = dict.keys()
+    keys.sort()
+    for key in keys:
+        value = dict[key]
+        if type(value) in [list, tuple] :
+            if type(value) is tuple:
+                value = list(value)
+            value.sort()
+            value = ' '.join(value)
+        file.write('{0}\t{1}\n'.format(key, value))
+    file.close()
+
+
+# This function returns only the isotropic noises according to the specified RIR id
+# Please refer to ParseNoiseList() for the format of iso_noise_list
+def FilterIsotropicNoiseList(iso_noise_list, rir_id):
+    filtered_list = []
+    for noise in iso_noise_list:
+        if noise.rir_id == rir_id:
+            filtered_list.append(noise)
+
+    return filtered_list
+
+
+def AddPointSourceNoise(noise_addition_descriptor,  # descriptor to store the information of the noise added
+                        room,  # the room selected
+                        point_noise_list, # the point source noise list
+                        pointsource_noise_addition_probability, # Probability of adding point-source noises
+                        foreground_snrs, # the SNR for adding the foreground noises
+                        background_snrs, # the SNR for adding the background noises
+                        speech_dur,  # duration of the recording
+                        max_noises_recording  # Maximum number of point-source noises that can be added
+                        ):
+    if len(point_noise_list) > 0 and random.random() < pointsource_noise_addition_probability:
+        for k in range(random.randint(1, max_noises_recording)):
+            # pick the RIR to reverberate the point-source noise
+            noise = PickItemWithProbability(point_noise_list)
+            noise_rir = PickItemWithProbability(room.rir_list)
+            if noise.bg_fg_type == "background":
```

Comment:
Clearly mention how background and foreground noises are treated differently here.